src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,208 @@
+begin
+  haveI := is_iso.id X,
+  exact iso_has_right_lifting_property i (category_theory.as_iso (𝟙 X)),
+end
+
+/- An equivalent characterization for right lifting against a map i whose source is initial. 
+∅ → X
+↓   ↓
+B → Y has a lifting iff there is a map B → X making the right part commute.
+-/
+lemma right_lifting_property_initial_iff (i p : arrow C) 
+(h : category_theory.limits.is_initial i.left) :
+has_lifting_property i p ↔ ∀ {e : i.right ⟶ p.right}, ∃ l : i.right ⟶ p.left, l ≫ p.hom = e :=
+begin
+  split,
+  { intros hlift bottom,
+    have comm : (category_theory.limits.is_initial.to h p.left) ≫ p.hom = i.hom ≫ bottom := 
+    begin apply category_theory.limits.is_initial.hom_ext h, end,
+    haveI := hlift (arrow.hom_mk comm),
+    use (arrow.has_lift.struct (arrow.hom_mk comm)).lift,
+    rw (arrow.has_lift.struct (arrow.hom_mk comm)).fac_right,
+    refl },
+  { intros h1 sq,
+    cases h1 with e he,
+    fconstructor,
+    fconstructor,
+    fconstructor,
+    { exact e },
+    { apply category_theory.limits.is_initial.hom_ext, simp, exact h },
+    { exact he } }
+end
+
+/- A helper construction: given a square between i and f ≫ g, produce a square between i and g, 
+whose top leg uses f:
+A  → X  
+     ↓f
+↓i   Y             --> A → Y 
+     ↓g                ↓i  ↓g
+B  → Z                 B → Z 
+ -/
+def square_to_second {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z} (sq : i ⟶ arrow.mk (f ≫ g)) :
+i ⟶ arrow.mk g :=
+{ left := sq.left ≫ f,
+  right := sq.right,
+  w' := begin
+    have : sq.left ≫ f ≫ g = i.hom ≫ sq.right := arrow.w sq,
+    rw ←category.assoc at this,
+    tidy,
+  end
+}
+
+/- The condition of having the rlp with respect to a morphism i is stable under composition-/
+lemma has_right_lifting_property_comp {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z}
+(hf : has_lifting_property i (arrow.mk f))
+(hg : has_lifting_property i (arrow.mk g)) : 
+has_lifting_property i (arrow.mk (f ≫ g)) :=
+begin
+  intro sq0, -- a square between i and f ≫ g
+  let sq1 := square_to_second sq0, -- transform this into a square between i and g
+  
+  -- lift i vs. g
+  haveI := hg sq1,
+  have lift_structure_sq1 := arrow.has_lift.struct sq1,
+  have sq1_lift := arrow.lift sq1,
+
+  -- form a square from i to f, using the previously constructed lift 
+  have h3 : sq0.left ≫ (arrow.mk f).hom = i.hom ≫ lift_structure_sq1.lift :=
+  begin
+    rw lift_structure_sq1.fac_left,
+    refl,
+  end,
+
+  -- construct a square i ⟶ f
+  let sq2 : i ⟶ (arrow.mk f) :=
+  {
+    left := sq0.left,
+    right := lift_structure_sq1.lift,
+    w' := begin
+      simp,
+      tidy,
+    end,
+  },
+  
+  -- construct a lift i vs. f
+  haveI := hf sq2,
+  have lift_structure_sq2 := arrow.has_lift.struct sq2,
+  have sq2_lift := arrow.lift sq2,
+  
+  -- show that this lift is a lift of i vs. g ∘ f
+  fconstructor,
+  fconstructor,
+  fconstructor,
+  { simp, exact lift_structure_sq2.lift, },
```

Comment:
Here you use `simp` on data, which is in general a bad idea. It (somewhat miraculously) works out in this case because the simp two lines below somehow cleans everything right up, but this proof should be refactored to avoid this.

More generally, we try to put `simp` (and things that call it, like `tidy`) only at the end of proofs. This makes proofs less likely to break when the set of `simp` lemmas changes. In the middle of proofs, it is generally a good idea to replace a `simp` by the output of `squeeze_simp`.

Reply:
OK, this chunk has now been replaced by some code that Scott suggested. Will keep it in mind though.